### PR TITLE
Reshuffle

### PR DESCRIPTION
**Pull request by cnegre**
I added a reshuffling routine at the end of the gpmdcov.f90 file.

### CODE REVIEW

File: examples/gpmdcov/gpmdcov.F90 [hunk 1]
```diff
@@ -1548,4 +1572,71 @@ subroutine gpmd_restart()
...
+      allocate(PartsInRankI(getNRanks()))
+    endif
+
+
+
+    reshuffle = 0
+    icount = 0
+    PartsInRankI = 0
+
+    do j=1,maxnparts
```
Comment (nicolasbock) [inline examples/gpmdcov/gpmdcov.F90]: When you reshuffle, you are not taking the cost per partition into account. How will this approach balance the load across the ranks?

File: examples/gpmdcov/gpmdcov.F90 [hunk 2]
```diff
@@ -309,13 +309,17 @@ subroutine gpmd_Part
       if(lt%verbose >= 1 .and. myRank == 1) write(*,*) "In prg_get_covgraph_h .."
       mls_ii = mls()
       call prg_get_covgraph_h(sy,nl%nnStructMindist,nl%nnStruct,nl%nrnnstruct,gsp2%nlgcut,graph_h,gsp2%mdim,lt%verbose)
-      if(lt%verbose >= 1 .and. myRank == 1) write(*,*) "Time for prg_get_covgraph_h "//to_string(mls()-mls_ii)//" ms"
+      if(lt%verbose >= 1 .and. myRank == 1) write(*,*) "Time for prg_get_covgraph_h ",(mls()-mls_ii)," ms"
```
Comment (cnegre) [inline examples/gpmdcov/gpmdcov.F90]: I changed it back now. I changed this for compatibility with the april 6 version.

File: examples/gpmdcov/gpmdcov.F90 [hunk 3]
```diff
@@ -309,13 +309,17 @@ subroutine gpmd_Part
...
 
 #ifdef DO_MPI
-      do ipt = gpat%localPartMin(myRank), gpat%localPartMax(myRank)
+       !do ipt= gpat%localPartMin(myRank), gpat%localPartMax(myRank)
+       do iptt=1,PartsInRankI(myRank)
+          ipt= reshuffle(iptt,myRank)
```
Comment (cnegre) [inline examples/gpmdcov/gpmdcov.F90]: Yes:
  !! PartsInRankI(getNRanks()) stores the number of partitions assigned to rank i
  !! reshuffle(j,i) assigns partition reshuffle(j,i) to rank i for j=1,PartsInRankI(getNRanks())

File: examples/gpmdcov/gpmdcov.F90 [hunk 4]
```diff
@@ -1548,4 +1572,71 @@ subroutine gpmd_restart()
 
   end subroutine gpmd_restart
 
+  !> Reshuffle the parts
+  subroutine gpmd_reshuffle()
+
+    maxnparts = 0
```
Comment (cnegre) [inline examples/gpmdcov/gpmdcov.F90]: Done,

File: examples/gpmdcov/gpmdcov.F90 [hunk 5]
```diff
@@ -1548,4 +1572,71 @@ subroutine gpmd_restart()
...
+    if(allocated(reshuffle))then
+      deallocate(reshuffle)
+      deallocate(PartsInRankI)
+      allocate(reshuffle(maxnparts,getNRanks()))
+      allocate(PartsInRankI(getNRanks()))
+    else
+      allocate(reshuffle(maxnparts,getNRanks()))
+      allocate(PartsInRankI(getNRanks()))
+    endif
+
```
Comment (cnegre) [inline examples/gpmdcov/gpmdcov.F90]: Done

File: examples/gpmdcov/gpmdcov.F90 (hunk 1 above)
Comment (cnegre) [inline examples/gpmdcov/gpmdcov.F90]: The partitions are ordered in decreasing ch sizes (gpat%sgraph(ipt)%lsize). The reshuffling ensures that the load of the ranks get balanced.

Comment (nicolasbock) [inline examples/gpmdcov/gpmdcov.F90]: Ah, thanks

File: examples/gpmdcov/gpmdcov.F90 (hunk 3 above)
Comment (nicolasbock) [inline examples/gpmdcov/gpmdcov.F90]: Thanks

**Pull request merged**